### DISCUSSION TITLE
Need to delete the file from history

### DISCUSSION THREAD

**Discussion by zhussain-equinix** [community/community#56686] [Actions]
### Select Topic Area

General

### Body

Hi,

I have committed a  file, later i realize and deleted the file from the GitHub, but the file remain in the history.

Could you please let me know how can we delete the file from history.

Thanks
Zulfikar Hussain

**Comment by airtower-luna** [comment]
[Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) describes how to do that. Based on your description you'll probably have to edit the commit where you added the file, and then adjust (or remove) the one where you removed it. Look at [`git-filter-repo`](https://github.com/newren/git-filter-repo) if you're dealing with a complex history (with multiple branches containing the relevant commits, etc.), it's much easier than manually editing the history.

A few things to note:

* Editing history will change *all* commit IDs after the first removed (or changed) commit. This means anyone who already has a copy of your repository and tries to fetch updates will have to manually accept the modified history.
* If you want the file to be *completely gone* you'll need to contact GitHub support and ask them purge the relevant caches, see: [Remove sensitive data](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository)
* You should probably consider the data in the file compromised anyway and take appropriate action (e.g. revoke credentials, notify people who's private date was accidentally published, …).

**Comment by nirav-gajera** [comment]
1. Clone the repository to your local
` git clone <repository-url>
`

2. Use the filter-branch command to remove the file from the entire Git history. Replace <file-path> with the path to the file you want to delete  ex:- `git filter-branch --tree-filter 'rm -f <file-path>' HEAD
`

3.If the file is located in a specific directory, you would use: `git filter-branch --tree-filter 'rm -f <directory-path>/<file-name>' HEAD
`

4.Force push the updated history  : `git push --force origin master
`
